EVCC crashes when reading coils

### Describe the bug

I have a custom wallbox configured that reads coils to determine the status.
In the past the datatype given there was not relevant and I used to use bool8.
This datatype is no longer supported, so I used uint16 and also tried int16. Both leads to a crash.

It seems the modbus `BoolGetter()` has been reworked in the latest release. It now uses `FloatGetter()` which crashes on the data, which probably is a single byte.


### Steps to reproduce

1. Configure the charger as shown below
2. Start EVCC
3. Connect a Vehicle
...


### Configuration details

```yaml
The charger definition I use is:

chargers:
- type: custom
  name: wallbox_custom
  tos: true
  status:
    source: modbus
    id: 180
    uri: 192.168.100.8:502
    register:
      address: 100
      type: input
      decode: int16
  enabled:
    source: modbus
    id: 180
    uri: 192.168.100.8:502
    register:
      address: 400
      type: coil
      decode: int16
  enable:
    source: modbus
    id: 180
    uri: 192.168.100.8:502
    register:
      address: 400
      type: writecoil
      decode: int16
  maxcurrent:
    source: modbus
    id: 180
    uri: 192.168.100.8:502
    register:
      address: 300
      type: writeholding
      decode: uint16
  phases1p3p:
    source: script
    cmd: /etc/switch-phases.sh ${phases}
    timeout: 15s
```


### Log details

```text
karsten@evcc:~$ evcc -l debug
[main  ] INFO 2024/03/10 11:59:37 evcc 0.124.9
[main  ] INFO 2024/03/10 11:59:37 using config file: /etc/evcc.yaml
[main  ] INFO 2024/03/10 11:59:37 starting ui and api at :7070
[db    ] INFO 2024/03/10 11:59:37 using sqlite database: /home/karsten/.evcc/evcc.db
[mqtt  ] INFO 2024/03/10 11:59:37 connecting evcc-1145160712 at tcp://iobroker.home.lan:1885
[mqtt  ] DEBUG 2024/03/10 11:59:37 tcp://iobroker.home.lan:1885 connected
[lp-1  ] DEBUG 2024/03/10 11:59:40 charge total import: 761.844kWh
[lp-2  ] DEBUG 2024/03/10 11:59:40 charge total import: 0.000kWh
[site  ] WARN 2024/03/10 11:59:41 vehicle 'Roller' does not support automatic detection
[site  ] INFO 2024/03/10 11:59:41 site config:
[site  ] INFO 2024/03/10 11:59:41   meters:      grid ✓ pv ✓ battery ✓
[site  ] INFO 2024/03/10 11:59:41     grid:      power ✓ energy ✗ currents ✗
[site  ] INFO 2024/03/10 11:59:41     pv 1:      power ✓ energy ✗ currents ✗
[site  ] INFO 2024/03/10 11:59:41     battery 1: power ✓ energy ✗ currents ✗ soc ✓ capacity ✓
[site  ] INFO 2024/03/10 11:59:41     battery 2: power ✓ energy ✗ currents ✗ soc ✓ capacity ✓
[site  ] INFO 2024/03/10 11:59:41   vehicles:
[site  ] INFO 2024/03/10 11:59:41     vehicle 1: range ✓ finish ✓ status ✓ climate ✗ wakeup ✗
[site  ] INFO 2024/03/10 11:59:41     vehicle 2: range ✗ finish ✗ status ✗ climate ✗ wakeup ✗
[site  ] INFO 2024/03/10 11:59:41     vehicle 3: range ✓ finish ✓ status ✓ climate ✗ wakeup ✓
[lp-1  ] INFO 2024/03/10 11:59:41 loadpoint 1:
[lp-1  ] INFO 2024/03/10 11:59:41   mode:        pv
[lp-1  ] INFO 2024/03/10 11:59:41   charger:     power ✗ energy ✗ currents ✗ phases ✓ wakeup ✗
[lp-1  ] INFO 2024/03/10 11:59:41   meters:      charge ✓
[lp-1  ] INFO 2024/03/10 11:59:41     charge:    power ✓ energy ✓ currents ✓
[lp-2  ] INFO 2024/03/10 11:59:41 loadpoint 2:
[lp-2  ] INFO 2024/03/10 11:59:41   mode:        pv
[lp-2  ] INFO 2024/03/10 11:59:41   charger:     power ✓ energy ✓ currents ✗ phases ✗ wakeup ✗
[lp-2  ] INFO 2024/03/10 11:59:41   meters:      charge ✓
[lp-2  ] INFO 2024/03/10 11:59:41     charge:    power ✓ energy ✓ currents ✗
[lp-1  ] DEBUG 2024/03/10 11:59:41 phase timer inactive
[lp-1  ] DEBUG 2024/03/10 11:59:41 pv timer inactive
panic: runtime error: index out of range [1] with length 1

goroutine 1 [running]:
encoding/binary.bigEndian.Uint16(...)
        encoding/binary/binary.go:143
github.com/volkszaehler/mbmd/encoding.Int16({0xc00010e7e9?, 0xc0007d7920?, 0xc000c0cde0?})
        github.com/volkszaehler/mbmd@v0.0.0-20231215091549-af16b1f597b9/encoding/number.go:19 +0x2e
github.com/evcc-io/evcc/util/modbus.Register.DecodeFunc.asFloat64[...].func1()
        github.com/evcc-io/evcc/util/modbus/register.go:235 +0x1c
github.com/evcc-io/evcc/provider.(*Modbus).FloatGetter.func1()
        github.com/evcc-io/evcc/provider/modbus.go:116 +0x82
github.com/evcc-io/evcc/provider.(*Modbus).BoolGetter.func1()
        github.com/evcc-io/evcc/provider/modbus.go:158 +0x13
github.com/evcc-io/evcc/charger.(*Charger).Enabled(...)
        github.com/evcc-io/evcc/charger/charger.go:159
github.com/evcc-io/evcc/core.(*Loadpoint).Prepare(0xc001184c08, 0x0?, 0x52c8320?, 0xc00097d320?)
        github.com/evcc-io/evcc/core/loadpoint.go:639 +0xf35
github.com/evcc-io/evcc/core.(*Site).Prepare(0xc000bbfba0, 0xc000c8aa20, 0xc0016c2480)
        github.com/evcc-io/evcc/core/site.go:877 +0x17b
github.com/evcc-io/evcc/cmd.runRoot(0x522cb80, {0x28ff8d3?, 0x7?, 0x28f299b?})
        github.com/evcc-io/evcc/cmd/root.go:259 +0x124f
github.com/spf13/cobra.(*Command).execute(0x522cb80, {0xc000050640, 0x2, 0x2})
        github.com/spf13/cobra@v1.8.0/command.go:987 +0xab1
github.com/spf13/cobra.(*Command).ExecuteC(0x522cb80)
        github.com/spf13/cobra@v1.8.0/command.go:1115 +0x3ff
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.8.0/command.go:1039
github.com/evcc-io/evcc/cmd.Execute()
        github.com/evcc-io/evcc/cmd/root.go:100 +0x1a
main.main()
        github.com/evcc-io/evcc/main.go:44 +0x2d

```
```


### What type of operating system are you running?

Linux

### Version

0.124.9

--- a/provider/modbus.go
+++ b/provider/modbus.go
@@ -174,12 +174,20 @@ func (m *Modbus) writeFunc() (func(float64) error, error) {
 	return func(val float64) error {
 		val *= m.scale
 
-		switch {
-		case op.FuncCode == gridx.FuncCodeWriteSingleRegister:
+		switch op.FuncCode {
+		case gridx.FuncCodeWriteSingleCoil:
+			var uval uint16
+			if val != 0 {
+				uval = 0xFF00
+			}
+			_, err = m.conn.WriteSingleCoil(op.Addr, uval)
+			return err
+
+		case gridx.FuncCodeWriteSingleRegister:
 			_, err = m.conn.WriteSingleRegister(op.Addr, uint16(val))
 			return err
 
-		case op.FuncCode == gridx.FuncCodeWriteMultipleRegisters:
+		case gridx.FuncCodeWriteMultipleRegisters:
 			b, err := encode(val)
 			if err == nil {
 				_, err = m.conn.WriteMultipleRegisters(op.Addr, op.Length, b)

--- a/util/modbus/modbus.go
+++ b/util/modbus/modbus.go
@@ -182,8 +182,8 @@ func (mb *Connection) ReadCoils(address, quantity uint16) ([]byte, error) {
 	return mb.ReadCoilsWithSlave(mb.slaveID, address, quantity)
 }
 
-func (mb *Connection) WriteSingleCoil(address, quantity uint16) ([]byte, error) {
-	return mb.WriteSingleCoilWithSlave(mb.slaveID, address, quantity)
+func (mb *Connection) WriteSingleCoil(address, value uint16) ([]byte, error) {
+	return mb.WriteSingleCoilWithSlave(mb.slaveID, address, value)
 }
 
 func (mb *Connection) ReadInputRegisters(address, quantity uint16) ([]byte, error) {

--- a/util/modbus/register.go
+++ b/util/modbus/register.go
@@ -51,7 +51,7 @@ func (r Register) Length() (uint16, error) {
 	case strings.Contains(enc, "64"):
 		return 4, nil
 	default:
-		return 0, fmt.Errorf("invalid register encoding: %s", enc)
+		return 0, fmt.Errorf("invalid register length: %s", enc)
 	}
 }
 
@@ -171,6 +171,9 @@ func (r Register) EncodeFunc() (func(float64) ([]byte, error), error) {
 	enc := strings.ToLower(r.encoding())
 
 	switch {
+	case strings.HasPrefix(enc, "bool"):
+		fallthrough
+
 	case strings.HasPrefix(enc, "int") || strings.HasPrefix(enc, "uint"):
 		return r.encodeToBytes(func(v float64) uint64 {
 			return uint64(v)

